File: wiki/src/main/java/io/github/freya022/wiki/java/commands/slash/SlashBulkBan.java
package io.github.freya022.wiki.java.commands.slash;

import io.github.freya022.botcommands.api.commands.annotations.Command;
import io.github.freya022.botcommands.api.commands.application.ApplicationCommand;
import io.github.freya022.botcommands.api.commands.application.slash.GuildSlashEvent;
import io.github.freya022.botcommands.api.commands.application.slash.annotations.JDASlashCommand;
import io.github.freya022.botcommands.api.commands.application.slash.annotations.MentionsString;
import io.github.freya022.botcommands.api.commands.application.slash.annotations.SlashOption;
import io.github.freya022.botcommands.api.core.entities.InputUser;
import io.github.freya022.wiki.switches.wiki.WikiLanguage;
import net.dv8tion.jda.api.entities.IMentionable;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.entities.UserSnowflake;
import net.dv8tion.jda.api.exceptions.ErrorHandler;
import net.dv8tion.jda.api.requests.ErrorResponse;

import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;

//TODO remove wildcard once fixed
@SuppressWarnings("unchecked")
@WikiLanguage(WikiLanguage.Language.JAVA)
// --8<-- [start:bulk_ban-java]
@Command
public class SlashBulkBan extends ApplicationCommand {
    @JDASlashCommand(name = "bulk_ban", description = "Ban users in bulk")
    public void onSlashBulkBan(
            GuildSlashEvent event,
            @SlashOption(description = "Users to ban") @MentionsString List<? extends InputUser> users,
            @SlashOption(description = "Time frame of messages to delete") Long timeframe,
            @SlashOption(description = "Unit of the time frame", usePredefinedChoices = true) TimeUnit unit
    ) {
        // Check if any member cannot be banned
        final var higherMembers = new ArrayList<Member>();
        for (var user : users) {
            final Member member = user.getMember();
            if (member == null) continue;

            if (!event.getGuild().getSelfMember().canInteract(member)) {
                higherMembers.add(member);
            }
        }

        if (!higherMembers.isEmpty()) {
            final String mentions = higherMembers.stream().map(IMentionable::getAsMention).collect(Collectors.joining());
            event.reply("Cannot ban " + mentions + " as they have equal/higher roles")
                    .setEphemeral(true)
                    .queue();
            return;
        }

        event.deferReply(true).queue();

        event.getGuild().ban((List<UserSnowflake>) (List<? extends UserSnowflake>) users, Duration.of(timeframe, unit.toChronoUnit()))
                .queue(response -> {
                    event.getHook().sendMessageFormat("Banned %s users, %s failed", response.getBannedUsers().size(), response.getFailedUsers().size()).queue();
                }, new ErrorHandler()
                        .handle(ErrorResponse.MISSING_PERMISSIONS, exception -> {
                            event.getHook().sendMessage("Could not ban users due to missing permissions").queue();
                        })
                        .handle(ErrorResponse.FAILED_TO_BAN_USERS, exception -> {
                            event.getHook().sendMessage("Could not ban anyone").queue();
                        })
                );
    }
}
// --8<-- [end:bulk_ban-java]

File: wiki/src/main/kotlin/io/github/freya022/wiki/commands/slash/SlashBulkBan.kt
package io.github.freya022.wiki.commands.slash

import dev.minn.jda.ktx.coroutines.await
import dev.minn.jda.ktx.messages.reply_
import dev.minn.jda.ktx.messages.send
import io.github.freya022.botcommands.api.commands.annotations.Command
import io.github.freya022.botcommands.api.commands.application.ApplicationCommand
import io.github.freya022.botcommands.api.commands.application.slash.GuildSlashEvent
import io.github.freya022.botcommands.api.commands.application.slash.annotations.JDASlashCommand
import io.github.freya022.botcommands.api.commands.application.slash.annotations.MentionsString
import io.github.freya022.botcommands.api.commands.application.slash.annotations.SlashOption
import io.github.freya022.botcommands.api.core.entities.InputUser
import io.github.freya022.botcommands.api.core.utils.awaitCatching
import io.github.freya022.botcommands.api.core.utils.awaitUnit
import io.github.freya022.botcommands.api.core.utils.handle
import io.github.freya022.wiki.switches.wiki.WikiLanguage
import net.dv8tion.jda.api.entities.Guild
import net.dv8tion.jda.api.entities.UserSnowflake
import net.dv8tion.jda.api.requests.ErrorResponse
import java.util.concurrent.TimeUnit
import kotlin.time.Duration
import kotlin.time.toDuration
import kotlin.time.toDurationUnit
import kotlin.time.toJavaDuration

@WikiLanguage(WikiLanguage.Language.KOTLIN)
// --8<-- [start:bulk_ban-kotlin]
@Command
class SlashBulkBan : ApplicationCommand() {
    @JDASlashCommand(name = "bulk_ban", description = "Ban users in bulk")
    suspend fun onSlashBulkBan(
        event: GuildSlashEvent,
        @SlashOption(description = "Users to ban") @MentionsString users: List<InputUser>,
        @SlashOption(description = "Time frame of messages to delete") timeframe: Long,
        @SlashOption(description = "Unit of the time frame", usePredefinedChoices = true) unit: TimeUnit,
    ) {
        // Check if any member cannot be banned
        val higherMembers = users.mapNotNull { it.member }.filterNot { event.guild.selfMember.canInteract(it) }
        if (higherMembers.isNotEmpty()) {
            return event.reply_("Cannot ban ${higherMembers.joinToString { it.asMention }} as they have equal/higher roles", ephemeral = true).awaitUnit()
        }

        event.deferReply(true).queue()

        event.guild.ban(users, timeframe.toDuration(unit.toDurationUnit())).awaitCatching()
            // Make sure to use onSuccess first,
            // as 'handle' will clear the result type
            .onSuccess {
                event.hook.send("Banned ${it.bannedUsers.size} users, ${it.failedUsers.size} failed").await()
            }
            .handle(ErrorResponse.MISSING_PERMISSIONS) {
                event.hook.send("Could not ban users due to missing permissions").await()
            }
            .handle(ErrorResponse.FAILED_TO_BAN_USERS) {
                event.hook.send("Could not ban anyone").await()
            }
            // Throw unhandled exceptions
            .getOrThrow()
    }
}
// --8<-- [end:bulk_ban-kotlin]

private fun Guild.ban(users: Collection<UserSnowflake>, duration: Duration) =
    ban(users, duration.toJavaDuration())